### PR TITLE
Move DEM examples to documentation

### PR DESCRIPTION
**Pull request by shahabgol**
# Description of the problem

This PR moves the DEM examples to the new documentation.

### CODE REVIEW

File: doc/source/examples/dem/silo/silo.rst [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+
+Running the simulation
+----------------------
+This simulation can be launched by (in parallel mode on 32 processes):
+
+.. code-block:: text
+
+  mpirun -np 32 dem_3d silo_Golshan.prm
+
```
Comment (blaisb) [inline doc/source/examples/dem/silo/silo.rst:177]: Can you give an estimation of run time?

**Review by blaisb — COMMENTED** [review]
Very good work. Some stuff to address, but nothing major. I see tremendous efforts went in there that's for sure!

File: doc/source/examples/dem/rotating-drum/rotating-drum.rst [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        set rolling friction wall         	      	  	= 0.01
+    end
+
+
+Model parameters
+~~~~~~~~~~~~~~~~~
+
+Load-balancing updates the distribution of the subdomains between the processes in parallel simulation to achieve better computational performance (less simulation time). Three load-balancing methods are available in Lethe-DEM: ``once``, ``frequent``, or ``dynamic``. Read `this article <https://www.mdpi.com/2227-9717/10/1/79>`_ for more information about different load-balancing methods and their performances in various types of DEM simulations.
+
+Selecting ``repartition method = once``, requires defining the step at which the code calls load balancing (``load balance step``). ``Dynamic`` ``repartition method`` requires defining ``load balance frequency``, and in ``dynamic`` ``repartition method``, we should define ``load balance threshold`` and ``dynamic load balance check frequency``. In ``dynamic`` load balancing, the code checks the distribution of particles among the processors, every ``dynamic load balance check frequency`` steps, and if
```
Comment (tonielgeitani) [inline doc/source/examples/dem/rotating-drum/rotating-drum.rst:102]:
```suggestion
Selecting ``repartition method = once``, requires defining the step at which the code calls load balancing (``load balance step``). ``Frequent`` ``repartition method`` requires defining ``load balance frequency``, and in ``dynamic`` ``repartition method``, we should define ``load balance threshold`` and ``dynamic load balance check frequency``. In ``dynamic`` load balancing, the code checks the distribution of particles among the processors, every ``dynamic load balance check frequency`` steps, and if
```

Comment (tonielgeitani) [reply doc/source/examples/dem/rotating-drum/rotating-drum.rst:102]: Isn't the method that uses load balance frequency called frequent and not dynamic?

File: doc/source/examples/dem/rotating-drum/rotating-drum.rst [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+``fixed_wall`` is a static wall, and particles collide with these static walls upon reaching the wall. The only way to move these walls is to move the entire triangulation. If the ``outlet`` condition is chosen for a boundary, particles can leave the simulation domain via this outlet. Using ``rotational`` and ``translational`` boundary conditions, exerts imaginary rotational and translational velocities to that boundary. In other words, the boundary does not move, but the particles that have collisions with these walls receive a rotational or translational velocity from the wall. This feature is used in the rotating drum example. The boundary id of the ``cylinder`` side wall, defined with deal.ii grid generator is 4. We set the ``rotational speed`` equal to 11.6 rad/s, and the cylinder should rotate around its axis (`x` direction).
```
Comment (tonielgeitani) [inline doc/source/examples/dem/rotating-drum/rotating-drum.rst:130]:
```suggestion
``fixed_wall`` is a static wall, and particles collide with these static walls upon reaching them. The only way to move these walls is to move the entire triangulation. If the ``outlet`` condition is chosen for a boundary, particles can leave the simulation domain via this outlet. Using ``rotational`` or ``translational`` boundary conditions, exerts imaginary rotational and translational velocities to that boundary. In other words, the boundary does not move, but the particles that have collisions with these walls feel a rotational or translational velocity from the wall. This feature is used in the rotating drum example. The boundary id of the ``cylinder`` side wall, defined with deal.ii grid generator is 4. We set the ``rotational speed`` equal to 11.6 rad/s, and the cylinder should rotate around its axis (`x` direction).
```

File: doc/source/examples/dem/silo/silo.rst [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+This simulation consists of two stages: filling (0-4 s) and discharge (4-40 s) of particles. During the filling stage, we use a stopper (floating wall) to keep the inserted particles in the hopper region of the silo. When all the particles are inserted and packed in the hopper, we remove the stopper and particles leave the hopper.
+
+
+Parameter file
+--------------
+
+Mesh
+~~~~~
+
+Despite previous examples, in this example, we use a mesh generated using Gmsh. The Gmsh (with an extension of .msh) file is located inside the example folder.
```
Comment (tonielgeitani) [inline doc/source/examples/dem/silo/silo.rst:32]:
```suggestion
Contrary to previous examples, in this example, we use a mesh generated using Gmsh. The Gmsh (with an extension of .msh) file is located inside the example folder.
```

Comment (tonielgeitani) [reply doc/source/examples/dem/silo/silo.rst:32]: I don't think the word despite can be used in this context.

**Review by tonielgeitani — APPROVED** [review]

File: doc/source/examples/dem/dem.rst [hunk 5]
```diff
@@ -0,0 +1,11 @@
+****************************
+Discrete Element Method
+****************************
+
+We organize the DEM examples from the simplest (first example) to the most complicated example (last example). Example 1 (packing in circle), shows a two-dimensional simulation of particles packing in a circle. This example is extended in example 2 (packing in ball) to a three-dimensional simulation. Example 3 (rotating drum), shows a simulation in which the particles are packed in a cylinder and the cylinder is rotating). We simulate this rotation using a ``rotational`` boundary condition in Lethe-DEM. In example 4 (rotating drum with post-processing) we explain the post-processing features of Lethe-DEM using the same rotating drum simulation in example 3. In example 5 (rotation of box), we simulate the dynamics of particles in a rotating box. The key difference between examples 3 and 5 is that example 3 uses a ``rotational`` boundary condition to simulate the rotating drum, while example 5 uses a rotation of triangulation (``grid motion``). In example 6 (silo), we simulate the filling (using a ``floating wall``) and discharge of particles in a wedge-shaped silo.
```
[the remark this answers is not in the document]
Comment (shahabgol) [reply doc/source/examples/dem/dem.rst:5]: Done.

File: doc/source/examples/dem/packing-in-circle/packing-in-circle.rst [hunk 6]
```diff
@@ -0,0 +1,194 @@
+==================================
+Packing in Circle
+==================================
+
+This is the first example of Lethe-DEM. This example introduces the concept of parameter files to parametrize Lethe simulations. It is strongly recommended to visit `DEM parameters <https://lethe-cfd.github.io/lethe/parameters/dem/dem.html>`_ for more detailed information on the concepts and physical meanings of the parameters in Lethe-DEM.
```
[the remark this answers is not in the document]
Comment (shahabgol) [reply doc/source/examples/dem/packing-in-circle/packing-in-circle.rst:5]: Hyperlink from the documentation is used instead.

File: doc/source/examples/dem/packing-in-circle/packing-in-circle.rst [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        set insertion box minimum x            	 			= -0.05
+        set insertion box minimum y            	         	= 0
+        set insertion box maximum x            	         	= 0.05
+        set insertion box maximum y           	 		 	= 0.07
+        set insertion distance threshold					= 2
+        set insertion random number range					= 0.75
+        set insertion random number seed					= 19
+    end
+
+First, the ``insertion method`` is selected. There are two insertion methods (``uniform`` and ``non_uniform``) in Lethe-DEM. In ``uniform`` insertion the particles are inserted uniformly (without randomness in initial location), while in ``non_uniform``, particles are inserted randomly in the insertion box. ``inserted number of particles at each time step`` specifies the desired number of particles to be inserted at each insertion step.
```
[the remark this answers is not in the document]
Comment (shahabgol) [reply doc/source/examples/dem/packing-in-circle/packing-in-circle.rst:91]: Added in a note.

File: doc/source/examples/dem/packing-in-circle/packing-in-circle.rst [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+    subsection model parameters
+      set contact detection method 		   	   		= dynamic
+      set dynamic contact search size coefficient	= 0.7
+      set neighborhood threshold				 	= 1.5
+      set particle particle contact force method    = hertz_mindlin_limit_overlap
+      set particle wall contact force method        = nonlinear
+      set integration method				 		= velocity_verlet
+    end
+
+These parameters include ``contact detection method`` and its subsequent information (``dynamic contact search size coefficient`` **or** ``contact detection frequency`` for ``dynamic`` **or** ``constant`` contact detection method), ``neighborhood threshold`` (which defines the contact neighbor list size: ``neighborhood threshold`` * particle diameter), ``particle particle contact force method``, ``particle wall contact force method`` and ``integration method``. All the concepts, models and choices are explained in `DEM parameters <https://lethe-cfd.github.io/lethe/parameters/dem/dem.html>`_.
```
[the remark this answers is not in the document]
Comment (shahabgol) [reply doc/source/examples/dem/packing-in-circle/packing-in-circle.rst:155]: Dynamic contact search only needs one parameter (dynamic contact search size coefficient). Constant contact search also only needs one parameter (contact detection frequency). The theory of constant and dynamic contact search algorithms are added.

File: doc/source/examples/dem/rotating-drum/rotating-drum.rst [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+      set output frequency            		 = 1000
+    end
+
+Running the simulation
+----------------------
+This simulation can be launched (in parallel mode on 64 processes) by:
+
+.. code-block:: text
+
+  mpirun -np 64 dem_3d rotating_drum.prm
```
[the remark this answers is not in the document]
Comment (shahabgol) [reply doc/source/examples/dem/rotating-drum/rotating-drum.rst:165]: I have added a warning to report the approximate simulation time. I will set up a smaller case with a smaller number of particles and add it to this example. but I have to test it on my machine first. So I will do it in another PR.

File: doc/source/examples/dem/rotating-drum-with-post-processing/rotating-drum-with-post-processing.rst [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code-block:: text
+
+    subsection simulation control
+      set time step                 		 = 1e-6
+      set time end       					 = 15
+      set log frequency				         = 1000
+      set output frequency            		 = 1000
+    end
```
[the remark this answers is not in the document]
Comment (shahabgol) [reply doc/source/examples/dem/rotating-drum-with-post-processing/rotating-drum-with-post-processing.rst:137]: done.

File: doc/source/examples/dem/rotating-drum-with-post-processing/rotating-drum-with-post-processing.rst [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+
+
+Running the simulation
+----------------------
+This simulation can be launched (in parallel mode on 64 processes) by:
+
+.. code-block:: text
+
+  mpirun -np 64 dem_3d rotating_drum_with_post_processing.prm
```
[the remark this answers is not in the document]
Comment (shahabgol) [reply doc/source/examples/dem/rotating-drum-with-post-processing/rotating-drum-with-post-processing.rst:165]: done.

File: doc/source/examples/dem/silo/silo.rst (hunk 1 above)
Comment (shahabgol) [reply doc/source/examples/dem/silo/silo.rst:177]: added.

File: doc/source/examples/dem/packing-in-ball/packing-in-ball.rst [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+  dem_3d packing_in_ball.prm
+
+We can also launch this simulation in parallel mode. For example, to launch the simulation on 8 processes:
+
+.. code-block:: text
+
+  mpirun -np 8 dem_3d packing_in_ball.prm
+
+.. note::
+	The parallel simulations are generally faster than simulations on a single process. However, to leverage the full performance of a parallel simulation, it should be performed with a load-balancing strategy throughout the simulation. Load-balancing is explained in the next example.
```
[the remark this answers is not in the document]
Comment (shahabgol) [reply doc/source/examples/dem/packing-in-ball/packing-in-ball.rst:144]: Added as a hyperlink.

File: doc/source/examples/dem/rotating-drum/rotating-drum.rst (hunk 2 above)
Comment (shahabgol) [reply doc/source/examples/dem/rotating-drum/rotating-drum.rst:102]: Absolutely right, I made a mistake here. Corrected.

File: doc/source/examples/dem/silo/silo.rst (hunk 4 above)
Comment (shahabgol) [reply doc/source/examples/dem/silo/silo.rst:32]: Corrected.

**Comment by shahabgol** [comment]
@blaisb @tonielgeitani I have addressed your comments in the last commit. Only one comment remains on designing a new example for rotating drum with a smaller number of particles. I will do it in a separate PR, as I need to have access to my workstation. :)

**Pull request merged**